source/common/router/router.cc

```diff
@@ -780,5 +765,37 @@ ProdFilter::createRetryState(const RetryPolicy& policy, Http::HeaderMap& request
                                 priority);
 }
 
+void Filter::UpstreamRequest::setRequestEncoder(Http::StreamEncoder& request_encoder) {
+  request_encoder_ = &request_encoder;
+  // Now that there is an encoder, have the connection manager inform the manager when the
+  // downstream buffers are overrun.  This may result in immediate watermark callbacks referencing
+  // the encoder.
+  parent_.callbacks_->addDownstreamWatermarkCallbacks(downstream_watermark_manager_);
+}
+
+void Filter::UpstreamRequest::clearRequestEncoder() {
+  // Before clearing the encoder, unsubscribe from callbacks.  This may result
+  // in some clean up calls to onBelowWriteBufferLowWatermark which reference
+  // the encoder.
+  if (request_encoder_) {
+    parent_.callbacks_->removeDownstreamWatermarkCallbacks(downstream_watermark_manager_);
+  }
+  request_encoder_ = nullptr;
+}
+
+void Filter::UpstreamRequest::DownstreamWatermarkManager::onAboveWriteBufferHighWatermark() {
+  ASSERT(parent_.request_encoder_);
+  // The downstream connection is overrun.  Pause reads from upstream.
+  parent_.parent_.cluster_->stats().upstream_flow_control_paused_reading_total_.inc();
+  parent_.request_encoder_->getStream().readDisable(true);
+}
+
+void Filter::UpstreamRequest::DownstreamWatermarkManager::onBelowWriteBufferLowWatermark() {
+  ASSERT(parent_.request_encoder_);
+  // The downstream connection has buffer available.  Resume reads from upstream.
+  parent_.parent_.cluster_->stats().upstream_flow_control_resumed_reading_total_.inc();
+  parent_.request_encoder_->getStream().readDisable(false);
```

Comment:
So the current version you're looking at no longer calls readDisable on teardown.  I instead updated the docs to say that it should be handled in the codec.

Checking out a prior version it's super duper hard to verify this in a unit test.    The codec test has this thing where all data is flushed immediately so if you readEnable the peer you end up under the stack of the framer twice and it's not reentrant and things crash with very confusing traces (not that I spent 6h debugging that one week).  The router doens't have actual buffer in these tests and I don't trust it to have real communication with the codec.

What I did was check out the old code where we readEnable on reset, stuck a boolean guard set true during runResetCallbacks and verified my integration test occasionally consumes  data under the stack of runResetCallbacks (it does) and with rc == 0 (it does).  That doesn't mean every type of disconnect and reset would work, but it does show that that particular one does.